pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionMetaDataTopicTailer.java

```diff
@@ -63,12 +65,22 @@ public void start() {
 
     @Override
     public void run() {
-        while(running) {
+        while (running) {
+            if (stopOnNoMessageAvailable) {
+                try {
+                    if (!reader.hasMessageAvailable()) {
+                        break;
+                    }
+                } catch (PulsarClientException e) {
+                    log.error("Received exception while testing hasMessageAvailable", e);
+                    errorNotifier.triggerError(e);
+                }
+            }
             try {
                 Message<byte[]> msg = reader.readNext();
                 processRequest(msg);
             } catch (Throwable th) {
-                if (running) {
+                if (running && !stopOnNoMessageAvailable) {
```

Comment:
I don't think we should check "! stopOnNoMessageAvailable". If errors occur while stopOnNoMessageAvailable=true, we still need to correctly handle the exception and not ignore it because that means we didn't process all messages successfully

Reply:
changed